### ISSUE TITLE
Feature request: support bulk ops for MultiMap, and TransactionalMultiMap

### ISSUE THREAD

**Issue by dmitrymz** [hazelcast/hazelcast#9079]
It will be great to add support for bulk operations for MultiMap, and TransactionalMultiMap, same as IMap.getAll(Set<K> keys), and putAll().

Absence of this functionality makes us use IMap with lists which is inefficient and high-maintenance

**Comment by burakcelebi** [comment]
Related: https://github.com/hazelcast/hazelcast/issues/46

**Comment by abdulazizali77** [comment]
Hi @mmedenjak @dmitrymz can i pick this up?

Just to confirm, i see that in addition to `getAll`, `putAll`, `IMap` has other bulk like interfaces `loadAll`, `removeAll`, `evictAll`; am guessing that `load` and `evict` might not be applicable to `MultiMap` and `TransactionalMultiMap`, but `remove` might be?
Additionally i see that `MultiMap` has a `delete` interface, should there be a `deleteAll` interface? So, what should the new bulk interfaces for `MultiMap` and `TransactionalMultiMap` be?

Also, maybe getting ahead of myself, but am inclined to have separate PRs for `MultiMap` and `TransactionalMultiMap` wdyt?

**Comment by mmedenjak** [comment]
Hi @abdulazizali77 ! 

Let's not rush it and let's first add `getAll` and `putAll`. In addition to that, can you make those methods async? That will allow users to wait for those methods to complete asynchronously instead of blocking the user thread until the invocations return. So, instead of `getAll` and `putAll`, you can add `getAllAsync` and `putAllAsync`. An example of a bulk and async method would be `MapProxyImpl#putAllAsync` and `ClientMapProxy#putAllAsync`. These methods are not yet public API but they will be soon. You can look at those methods for inspiration - the MultiMap versions won't be the same but will be similar. You'll also need to modify the [client protocol repo](https://github.com/hazelcast/hazelcast-client-protocol) to support these methods from hazelcast clients.

And yes, it's best if you send separate PRs for `MultiMap` and `TransactionalMultiMap`.

**Comment by abdulazizali77** [comment]
Thanks @mmedenjak ! ill get cracking on the `MultiMap` `getAllAsync` and `putAllAsync` first. 

Probably getting ahead of myself again, but Im assuming that for this work, ill probably only make changes to the `MultiMap.yaml` and `TransactionalMultiMap.yaml` definitions after both `MultiMap` and `TransactionalMultiMap` changes get in? wdyt?

**Comment by mmedenjak** [comment]
Yes, I think you'll have to add the definitions in the yaml files, re-generate the codecs as explained in the README document in your hazelcast repo branch and then use those new codecs in your hazelcast repo PR to add the client support.
For instance, you'll get a new codec generated like `MapFetchKeysCodec` and then you'll write a new task like `MapFetchKeysMessageTask` and then [add that task in `DefaultMessageTaskFactoryProvider` ](https://github.com/hazelcast/hazelcast/blob/04a303e00d3d17b0bf9584fc04ba152a54760ed0/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/DefaultMessageTaskFactoryProvider.java#L1408-L1409) and then invoke the task [like this](https://github.com/hazelcast/hazelcast/blob/97ff57ba92dc1fbfa77720885e7db40757999080/hazelcast/src/main/java/com/hazelcast/client/map/impl/iterator/ClientMapPartitionIterator.java#L72-L83).

Feel free to take a look at some other and simpler methods in `ClientMapProxy` to see how tasks and codecs interact.

**Comment by abdulazizali77** [comment]
Thanks @mmedenjak Ill get back to you should i encounter uncertainties!

**Comment by abdulazizali77** [comment]
Hi @mmedenjak apologies for the very late update, ive put up the initial PR and my thought about the design (and uncertainties) and implementation detailed there. PTAL

**Comment by abdulazizali77** [comment]
@mmedenjak , should the issue be renamed?

**Comment by mmedenjak** [comment]
No need, I'll just mention here that we will not be investing time into transactions at the moment. For more information, see [here](https://github.com/hazelcast/hazelcast/issues/9080#issuecomment-616559316). Users trying to create enhancement issues for transactional map can now still find this issue and see what was the discussion outcome.

**Comment by c-attila** [comment]
Hi! Does anyone know why was this not implemented in the end?
We would need this feature in our project using Hazelcast 5 Enterprise. A QueryCache for MultiMaps would be also good.
FYI @dmitrymz, @abdulazizali77, @mmedenjak
(see also #3840 still open, #16973 and #46 closed but related)
